### PR TITLE
Faster, better and differentiable Pauli decompose

### PR DESCRIPTION
**Pull request by obliviateandsurrender**
**Context:** Adds a Pauli decompose function that works with non-Hermitian matrices and has a differentiable workflow.

**Description of the Change:** Changes `pauli_decompose` method and adds relevant additional tests

**Benefits:** Faster decomposition and removes restriction for matrix being square or Hermitian. 

**Possible Drawbacks:** N/A

**Related GitHub Issues:**

### CODE REVIEW

**Review by trbromley — COMMENTED** [review]
Thanks @obliviateandsurrender, can this implementation be made differentiable?

**Comment by obliviateandsurrender** [comment]
@trbromley current version seems to be differentiable (at least with `autograd`).

**Comment by trbromley** [comment]
Thanks @obliviateandsurrender!

**Comment by obliviateandsurrender** [comment]
Some benchmarks:
image
image

File: tests/pauli/test_conversion.py [hunk 1]
```diff
@@ -28,25 +28,44 @@
...
         ):
             qml.pauli_decompose(hamiltonian)
 
-    def test_not_hermitian(self):
-        """Tests that an exception is raised if the Hamiltonian is not Hermitian, i.e.
-        equal to its own conjugate transpose"""
-        with pytest.raises(ValueError, match="The matrix is not Hermitian"):
-            qml.pauli_decompose(np.array([[1, 2], [3, 4]]))
+    @pytest.mark.parametrize("hamiltonian", [np.ones((5, 5)), np.ones((3, 3))])
+    def test_wrong_shape_non_power_two(self, hamiltonian):
```
Comment (soranjh) [inline tests/pauli/test_conversion.py:61]: We were not checking for this before?

File: tests/pauli/test_conversion.py [hunk 2]
```diff
@@ -107,7 +126,44 @@ def test_to_paulisentence(self, hamiltonian):
         assert isinstance(ps, qml.pauli.PauliSentence)
         assert np.allclose(hamiltonian, ps.to_mat(range(num_qubits)))
 
+    @pytest.mark.parametrize("hide_identity", [True, False])
+    @pytest.mark.parametrize("matrix", test_general_matrix)
+    def test_observable_types_general(self, matrix, hide_identity):
```
Comment (soranjh) [inline tests/pauli/test_conversion.py:131]: Why this new test is needed? The previous tests were not comprehensive enough?

Comment (obliviateandsurrender) [reply tests/pauli/test_conversion.py:131]: The previous functionality only worked with Hermitian matrices, but the current one works with more general matrices. New tests have been added for that purpose.

File: tests/pauli/test_conversion.py (hunk 1 above)
Comment (obliviateandsurrender) [reply tests/pauli/test_conversion.py:61]: There was a previous function called `test_wrong_shape`, but I had to rewrite it to support including the message for `padding`. So rewrote the tests as well.

File: pennylane/pauli/conversion.py [hunk 3]
```diff
@@ -94,52 +101,119 @@ def pauli_decompose(
...
+    for idx in range(num_qubits):
+        index = [slice(None)] * (2 * num_qubits)
+        index[idx] = index[idx + num_qubits] = 1
+        phase_mat[tuple(index)] *= 1j
+    phase_mat = qml.math.convert_like(qml.math.reshape(phase_mat, shape), matrix)
+
+    # c_00 + c_11 -> I; c_00 - c_11 -> Z; c_01 + c_10 -> X; 1j*(c_10 - c_01) -> Y
+    term_mat = qml.math.transpose(qml.math.multiply(hadamard_transform_mat, phase_mat))
+
+    # Convert to Hamiltonian and PauliSentence
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/pauli/conversion.py:188]: That's pretty much what we are doing in the rest of the code. Building `coeff` and `terms` for making `Hamiltonians` and `PauliSentences`. I can rewrite it as that to make it more explicit.

File: pennylane/pauli/conversion.py [hunk 4]
```diff
@@ -19,36 +19,43 @@
...
 def pauli_decompose(
-    H, hide_identity=False, wire_order=None, pauli=False
+    matrix,
+    hide_identity=False,
+    wire_order=None,
+    pauli=False,
+    padding=False,
 ) -> Union[Hamiltonian, PauliSentence]:
-    r"""Decomposes a Hermitian matrix into a linear combination of Pauli operators.
+    r"""Decomposes a matrix into a linear combination of Pauli operators acting on `n` qubits.
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/pauli/conversion.py:39]: I added it because we were not mentioning anything about `n` before we specify it in the shape `(2^n, 2^n)`.

File: pennylane/pauli/conversion.py [hunk 5]
```diff
@@ -94,52 +101,119 @@ def pauli_decompose(
...
+    # Account for the phases from Y
+    phase_mat = qml.math.ones(shape, dtype=complex).reshape((2,) * (2 * num_qubits))
+    for idx in range(num_qubits):
+        index = [slice(None)] * (2 * num_qubits)
+        index[idx] = index[idx + num_qubits] = 1
+        phase_mat[tuple(index)] *= 1j
+    phase_mat = qml.math.convert_like(qml.math.reshape(phase_mat, shape), matrix)
+
+    # c_00 + c_11 -> I; c_00 - c_11 -> Z; c_01 + c_10 -> X; 1j*(c_10 - c_01) -> Y
+    term_mat = qml.math.transpose(qml.math.multiply(hadamard_transform_mat, phase_mat))
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/pauli/conversion.py]: Added to the docstring.

File: pennylane/pauli/conversion.py [hunk 6]
```diff
@@ -19,36 +19,43 @@
...
 
 
+# pylint: disable=too-many-branches
 def pauli_decompose(
-    H, hide_identity=False, wire_order=None, pauli=False
+    matrix,
+    hide_identity=False,
+    wire_order=None,
+    pauli=False,
+    padding=False,
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/pauli/conversion.py:37]: I would prefer keeping it to `False` in order to prevent scenarios where users mistakenly input a matrix of a non `(2^n, 2^n)` shape.

File: doc/releases/changelog-dev.md [hunk 7]
```diff
@@ -98,6 +101,7 @@
 * The experimental `DefaultQubit2` device now supports computing VJPs and JVPs using the adjoint method.
   [(#4374)](https://github.com/PennyLaneAI/pennylane/pull/4374)
 
+
```
Comment (Jaybsoni) [inline doc/releases/changelog-dev.md:104]:
```suggestion
```

File: pennylane/pauli/conversion.py [hunk 8]
```diff
@@ -19,36 +19,52 @@
...
 ) -> Union[Hamiltonian, PauliSentence]:
-    r"""Decomposes a Hermitian matrix into a linear combination of Pauli operators.
+    r"""Decomposes a matrix into a linear combination of Pauli operators acting on :math:`n` qubits.
+
+    This method converts any matrix to a weighted sum of Pauli words using Bell-basis measurements
+    in :math:`O(n 4^n)`. The input matrix is first padded with zeros if its dimensions are not
```
Comment (DSGuala) [inline pennylane/pauli/conversion.py:42]:
```suggestion
    in time :math:`O(n 4^n)`. The input matrix is first padded with zeros if its dimensions are not
```
Don't use this suggestion. But is this a time requirement on the order of O(n4^n)? Could be good to specify that.

File: pennylane/pauli/conversion.py [hunk 9]
```diff
@@ -19,36 +19,52 @@
...
+    :math:`2^n\times 2^n` and written as a quantum state in the computational basis following the
+    `channel-state duality <https://en.wikipedia.org/wiki/Channel-state_duality>`_.
+    A Bell basis transformation is then performed using the
+    `Walsh-Hadamard transform <https://en.wikipedia.org/wiki/Hadamard_transform>`_, after which
+    coefficients for each of the :math:`4^n` Pauli words are computed while accounting for the
+    phase from each ``PauliY`` term occuring in the word.
```
Comment (DSGuala) [inline pennylane/pauli/conversion.py:48]: I thought this section was difficult to follow. Is it just a difficult concept? Or could it be explained with shorter, more direct sentences?

File: pennylane/pauli/conversion.py [hunk 10]
```diff
@@ -19,36 +19,52 @@
...
 
     Args:
-        H (array[complex]): a Hermitian matrix of dimension :math:`2^n\times 2^n`.
+        matrix (tensor_like[complex]): any matrix M, the keyword argument ``padding=True``
+            should be provided if the dimension of M is not :math:`2^n\times 2^n`.
```
Comment (DSGuala) [inline pennylane/pauli/conversion.py:52]: :cookie:

File: pennylane/pauli/conversion.py [hunk 11]
```diff
@@ -19,36 +19,52 @@
...
         hide_identity (bool): does not include the Identity observable within
             the tensor products of the decomposition if ``True``.
         wire_order (list[Union[int, str]]): the ordered list of wires with respect
             to which the operator is represented as a matrix.
         pauli (bool): return a PauliSentence instance if ``True``.
+        padding (bool): makes it compatible with rectangular matrices and square matrices
+            that are not of shape :math:`2^n\times 2^n` by padding them with zeros if ``True``.
```
Comment (DSGuala) [inline pennylane/pauli/conversion.py:59]:
```suggestion
        padding (bool): makes the function compatible with rectangular matrices and square matrices
            that are not of shape :math:`2^n\times 2^n` by padding them with zeros if ``True``.
```

File: pennylane/pauli/conversion.py [hunk 12]
```diff
@@ -19,36 +19,52 @@
...
 
     Returns:
         Union[~.Hamiltonian, ~.PauliSentence]: the matrix decomposed as a linear combination
         of Pauli operators, either as a :class:`~.Hamiltonian` or :class:`~.PauliSentence` instance.
 
     **Example:**
 
-    We can use this function to compute the Pauli operator decomposition of an arbitrary Hermitian
-    matrix:
+    We can use this function to compute the Pauli operator decomposition of an arbitrary matrix:
```
Comment (DSGuala) [inline pennylane/pauli/conversion.py:67]: :star_struck:

File: pennylane/pauli/conversion.py [hunk 13]
```diff
@@ -94,52 +110,121 @@ def pauli_decompose(
...
-        raise ValueError(
-            f"number of wires {len(wire_order)} is not compatible with number of qubits {n}"
-        )
+    .. details::
+        :title: Usage Details
+        :href: usage-decompose-operation
 
-    if wire_order is None:
-        wire_order = range(n)
+        For non-square matrices, we need to provide ``padding=True`` keyword argument:
```
Comment (DSGuala) [inline pennylane/pauli/conversion.py:118]:
```suggestion
        For non-square matrices, we need to provide the ``padding=True`` keyword argument:
```

File: pennylane/pauli/conversion.py [hunk 14]
```diff
@@ -94,52 +110,121 @@ def pauli_decompose(
...
+
+        >>> A = np.array([[-2, -2 + 1j]])
+        >>> H = qml.pauli_decompose(A, padding=True)
+        >>> print(H)
+          ((-1+0j)) [I0]
+        + ((-1+0.5j)) [X0]
+        + ((-1+0j)) [Z0]
+        + ((-0.5-1j)) [Y0]
+
+        We can also use the method within a differentiable workflow and obtain gradients:
```
Comment (DSGuala) [inline pennylane/pauli/conversion.py:128]: :rocket:

**Review by DSGuala — COMMENTED** [review]
Looks good. I left some suggestions/comments. As soon as those are addressed and I can stop getting the `ImportError: autoray couldn't find function 'gather' for backend 'builtins'.` and actually test the functionality, I will approve :sweat_smile:

File: pennylane/pauli/conversion.py (hunk 8 above)
Comment (obliviateandsurrender) [reply pennylane/pauli/conversion.py]: Yes, this is time complexity. Mentioned it.

File: pennylane/pauli/conversion.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/pauli/conversion.py:37]: Then the error message for `if shape[0] != 2**num_qubits:` should be modified to inform users that they can use 'padding=True'?

**Review by soranjh — APPROVED** [review]

File: pennylane/pauli/conversion.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/pauli/conversion.py:37]: Updated!

File: pennylane/pauli/conversion.py [hunk 15]
```diff
@@ -19,36 +19,52 @@
...
+    r"""Decomposes a matrix into a linear combination of Pauli operators.
+
+    This method converts any matrix to a weighted sum of Pauli words acting on :math:`n` qubits
+    in time :math:`O(n 4^n)`. The input matrix is first padded with zeros if its dimensions are not
+    :math:`2^n\times 2^n` and written as a quantum state in the computational basis following the
+    `channel-state duality <https://en.wikipedia.org/wiki/Channel-state_duality>`_.
+    A Bell basis transformation is then performed using the
+    `Walsh-Hadamard transform <https://en.wikipedia.org/wiki/Hadamard_transform>`_, after which
+    coefficients for each of the :math:`4^n` Pauli words are computed while accounting for the
+    phase from each ``PauliY`` term occuring in the word.
```
Comment (soranjh) [inline pennylane/pauli/conversion.py:48]: "occurring"

File: tests/pauli/test_conversion.py [hunk 16]
```diff
@@ -266,12 +363,6 @@ def test_operator(self, op, ps):
         """Test that PL arithmetic op is properly cast to a PauliSentence."""
         assert pauli_sentence(op) == ps
 
-    @pytest.mark.parametrize("op, ps", operator_ps)
-    def test_operator_private_ps(self, op, ps):
-        """Test that a correct pauli sentence is computed when passing an arithmetic operator and not
-        relying on the saved op._pauli_rep attribute."""
-        assert qml.pauli.conversion._pauli_sentence(op) == ps  # pylint: disable=protected-access
```
Comment (Jaybsoni) [inline tests/pauli/test_conversion.py:273]: Why was this removed ?

**Review by Jaybsoni — COMMENTED** [review]
Looks good overall! I still don't understand how this method works, but it seems to be performing well empirically. 

Left some comments, happy to review and approve once addressed

File: tests/pauli/test_conversion.py [hunk 17]
```diff
@@ -77,7 +96,7 @@ def test_observable_types(self, hamiltonian, hide_identity):
         allowed_obs = (Tensor, Identity, PauliX, PauliY, PauliZ)
 
         _, decomposed_obs = qml.pauli_decompose(hamiltonian, hide_identity).terms()
-        assert all(isinstance(o, allowed_obs) for o in decomposed_obs)
+        assert all((isinstance(o, allowed_obs) for o in decomposed_obs))
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply tests/pauli/test_conversion.py:99]: pylint was complaining about using generators iirc.

File: tests/pauli/test_conversion.py (hunk 16 above)
Comment (obliviateandsurrender) [reply tests/pauli/test_conversion.py:273]: I don't remember deleting these. It must be a mistake, let me restore them.

File: pennylane/pauli/conversion.py [hunk 18]
```diff
@@ -94,52 +110,123 @@ def pauli_decompose(
...
+            f"The matrix should be square, got {shape}. Use 'padding=True' for rectangular matrices."
+        )
 
-    obs_lst = []
-    coeffs = []
+    num_qubits = int(qml.math.log2(shape[0]))
+    if shape[0] != 2**num_qubits:
+        raise ValueError(
+            f"Dimension of the matrix should be a power of 2, got {shape}. Use 'padding=True' for these matrices."
+        )
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/pauli/conversion.py:165]: I would prefer not to do this by default for users not to do unintentional shape mismatches while defining their arrays. I guess it would make it difficult for them to debug their code.

File: pennylane/pauli/conversion.py [hunk 19]
```diff
@@ -94,52 +110,123 @@ def pauli_decompose(
...
+        + ((-1+0j)) [Z0]
+        + ((-0.5-1j)) [Y0]
+
+        We can also use the method within a differentiable workflow and obtain gradients:
+
+        >>> A = qml.numpy.array([[-2, -2 + 1j]], requires_grad=True)
+        >>> dev = qml.device("default.qubit", wires=1)
+        >>> @qml.qnode(dev)
+        ... def circuit(A):
+        ...    decomp = qml.pauli_decompose(A, padding=True)
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/pauli/conversion.py:134]: Nope works fine!
image

File: pennylane/pauli/conversion.py [hunk 20]
```diff
@@ -94,52 +110,123 @@ def pauli_decompose(
...
+    # Account for the phases from Y
+    phase_mat = qml.math.ones(shape, dtype=complex).reshape((2,) * (2 * num_qubits))
+    for idx in range(num_qubits):
+        index = [slice(None)] * (2 * num_qubits)
+        index[idx] = index[idx + num_qubits] = 1
+        phase_mat[tuple(index)] *= 1j
+    phase_mat = qml.math.convert_like(qml.math.reshape(phase_mat, shape), matrix)
+
+    # c_00 + c_11 -> I; c_00 - c_11 -> Z; c_01 + c_10 -> X; 1j*(c_10 - c_01) -> Y
+    # https://quantumcomputing.stackexchange.com/a/31790
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/pauli/conversion.py:200]: I just wanted to leave it as a reference for other devs.

File: pennylane/pauli/conversion.py [hunk 21]
```diff
@@ -94,52 +110,123 @@ def pauli_decompose(
...
+        indices.append(qml.math.bitwise_xor(indices[-1], (idx + 1) ^ (idx)))
+    term_mat = qml.math.cast(
+        qml.math.stack(
+            [qml.math.gather(matrix[idx], indice) for idx, indice in enumerate(indices)]
+        ),
+        complex,
+    )
+
+    # Perform Hadamard transformation on coloumns
+    hadamard_transform_mat = _walsh_hadamard_transform(qml.math.transpose(term_mat))
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/pauli/conversion.py:189]: good pylint.

File: pennylane/pauli/conversion.py [hunk 22]
```diff
@@ -19,36 +19,52 @@
...
-        H (array[complex]): a Hermitian matrix of dimension :math:`2^n\times 2^n`.
+        matrix (tensor_like[complex]): any matrix M, the keyword argument ``padding=True``
+            should be provided if the dimension of M is not :math:`2^n\times 2^n`.
         hide_identity (bool): does not include the Identity observable within
             the tensor products of the decomposition if ``True``.
         wire_order (list[Union[int, str]]): the ordered list of wires with respect
             to which the operator is represented as a matrix.
         pauli (bool): return a PauliSentence instance if ``True``.
+        padding (bool): makes the function compatible with rectangular matrices and square matrices
+            that are not of shape :math:`2^n\times 2^n` by padding them with zeros if ``True``.
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/pauli/conversion.py:59]: > Can we decompose a rectangle matrix without padding?
No, we cannot. 

I prefer keeping a `kwarg` to prevent scenarios where users mistakenly input a matrix of a non-standard shape. They would not know they are doing so unless we raise a `warning` in that case. I want them to be aware of what they are actually doing.

**Review by DSGuala — APPROVED** [review]
I checked the docs and tested a few examples with different square and non-square matrices.
Looks good to go! :shipit:

**Review by Jaybsoni — APPROVED** [review]
LGTM
